Documentation/design-docs/unloadability.md

```diff
@@ -0,0 +1,111 @@
+* The unload can be initiated eagerly using explicit `Unload` method or it can happen after all references to the `AssemblyLoadContext` and all the assemblies in it are gone.
+## Non-goals
+* Forceful unload using thread abort or similar measures to cause threads to stop running with frames of code from the context to be unloaded on the call stack.
+## General scenarios
+Based on various discussions and feedback on github, the following general scenarios were often mentioned as use cases for unloadability.
+* Loading assemblies for one-off introspection. 
+* Plugin scenarios when dynamic plugin loading and unloading is required.
+* Dynamically compiling, running and then flushing code. Useful for web sites, scripting engines, etc.
+## Real-world scenarios
+A couple of real-world scenarios were analyzed to assess feasibility of the `AssemblyLoadContext` unloading as a replacement of `AppDomain` unloading and also for possible new usages of the `AssemblyLoadContext` unloading.
+### Roslyn
+Roslyn has historically used AppDomains for executing different tests that can have assemblies with the same names but different identity or even the same identity, but different contents. `AppDomain` per test was very slow. So, to speed it up, they made them reusable based on the assemblies loaded into them.
+* In .NET Core, they use `AssemblyLoadContext` without unloading, one context per test works fine. 
+* Due to the missing plugin model, especially the dependency resolving issues, they use Mono.
+
+In future, the ability to use unloadable plugins in the compiler server would be great. Analyzers execute user code in the server, which is not good without the isolation. Ideally, whenever the compilation is executed with analyzers, it would run inside `AssemblyLoadContext`. The context would get unloaded after the build.
+### ASP.NET Core
+ASP.NET (not Core) originally used AppDomains to support dynamic compiling and running code behind pages. ASP.NET Core moved to a more static model because of the lack of ability to unload stuff. Many of their customers did the same thing too. So, there is no pressing need for unloadability there at the moment.
+
+However, they use two tool that could potentially benefit from the unloadability
+* Dotnet watch tool that watches a directory with sources and when a source file changes, it triggers recompilation and re-deployment. So, people can edit a source code used for their web page and the web server gets automatically recompiled and restarted. It is super-slow due to the process restart. So, the ability to unload and reload just the modified assemblies without process restart would be very helpful here. See https://docs.microsoft.com/en-us/aspnet/core/tutorials/dotnet-watch?view=aspnetcore-2.1 for more details on the watch tool usage in ASP.NET Core.
```

Comment:
It would not say it is super-slow. It is slower than the competing environments that do not have process restarts.